src/VisualStudio/Core/Def/Diagnostics/VisualStudioCodeAnalysisSuggestionsConfigService.AnalyzerConfigSummaryHelper.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeAnalysisSuggestions;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.CodeAnalysisSuggestions;
+
+using TreeOptions = ImmutableDictionary<string, ReportDiagnostic>;
+
+internal sealed partial class VisualStudioCodeAnalysisSuggestionsConfigService
+{
+    private record class AnalyzerConfigSummary(int WarningsAndErrorsCount, ImmutableHashSet<string> ConfiguredDiagnosticIds, string DiagnosticIdPattern);
+    private record class FirstPartyAnalyzerConfigSummary(AnalyzerConfigSummary CodeQualitySummary, AnalyzerConfigSummary CodeStyleSummary);
+
+    private static class AnalyzerConfigSummaryHelper
+    {
+        private static readonly ConditionalWeakTable<AnalyzerOptions, FirstPartyAnalyzerConfigSummary?> s_summaryCache = new();
+
+        public static FirstPartyAnalyzerConfigSummary? GetAnalyzerConfigSummary(Project project, IGlobalOptionService globalOptions)
+        {
+            if (!globalOptions.GetOption(CodeAnalysisSuggestionsOptionsStorage.ShowCodeAnalysisSuggestionsInLightbulb))
+                return null;
+
+            if (s_summaryCache.TryGetValue(project.AnalyzerOptions, out var summary))
+                return summary;
+
+            if (project.GetAnalyzerConfigOptions() is { } analyzerConfigOptions
+                && project.GetGlobalAnalyzerConfigOptions() is { } globalAnalyzerConfigOptions)
+            {
+                summary = CreateAnalyzerConfigSummary(analyzerConfigOptions.TreeOptions, globalAnalyzerConfigOptions.TreeOptions);
+            }
+            else
+            {
+                summary = null;
+            }
+
+            return s_summaryCache.GetValue(project.AnalyzerOptions, new(_ => summary));
+        }
+
+        private static FirstPartyAnalyzerConfigSummary? CreateAnalyzerConfigSummary(TreeOptions treeOptions, TreeOptions globalOptions)
+        {
+            const string CodeQualityPattern = "[cC][aA][0-9]{4}";
+            const string CodeStylePattern = "[iI][dD][eE][0-9]{4}";
```

Comment:
what are these?

Reply:
Regular expressions for CAxxxx and IDExxxx diagnostic IDs, added a comment